source/tutorials/file-sets.md

```diff
@@ -0,0 +1,661 @@
+
+```shell-session
+nix-repl> fs.trace ./.
+trace: /home/user (all files in directory)
+«lambda @ /nix/store/1czr278x24s3bl6qdnifpvm5z03wfi2p-nixpkgs-src/lib/fileset/default.nix:555:8»
+```
+:::
+
+Even though file sets conceptually contain local files, these files are *never* added to the Nix store unless explicitly requested.
+So You don't have to worry as much about accidentally copying secrets into the world-readable store.
+
+In this example, although we pretty-printed the home directory, no files were copied.
+This is in contrast to coercion of paths to strings such as in `"${./.}"`,
+which copies the whole directory to the Nix store on evaluation!
+
+:::{warning}
+When using [experimental Flakes](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-flake), local directories with a `flake.nix` are always copied into the Nix store *completely* unless it is a Git repository!
+:::
+
+This implicit coercion also works for files:
+
+```shell-session
+$ touch some-file
+```
+
+```shell-session
+nix-repl> fs.trace ./some-file
+trace: /home/user
+trace: - some-file (regular)
+```
+
+In addition to the included file, this also prints its [file type](https://nixos.org/manual/nix/stable/language/builtins.html#builtins-readFileType).
+
+
+## Example project
+
+To further experiment with the library, make a sample project.
+Create a new directory, enter it,
+and set up `niv` to pin the Nixpkgs dependency:
+
+```shell-session
+$ mkdir fileset
+$ cd fileset
+$ nix-shell -p niv --run "niv init --nixpkgs nixos/nixpkgs --nixpkgs-branch master"
+```
+
+Then create a `default.nix` file with the following contents:
+
+```{code-block} nix
+:caption: default.nix
+{
+  system ? builtins.currentSystem,
+  sources ? import ./nix/sources.nix,
+}:
+let
+  pkgs = import sources.nixpkgs {
+    config = { };
+    overlays = [ ];
+    inherit system;
+  };
+in
+pkgs.callPackage ./build.nix { }
+```
+
+Add two source files to work with:
+
+```shell-session
+$ echo hello > hello.txt
+$ echo world > world.txt
+```
+
+## Adding files to the Nix store
+
+To add files in a given file set to the Nix store,
+we use [`toSource`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.fileset.toSource).
+This function needs a `root` attribute to know which path to use as the root of the resulting store path,
+but only exactly the files in the `fileset` argument are included in the result.
+
+Define `build.nix` as follows:
+
+```{code-block} nix
+:caption: build.nix
+{ stdenv, lib }:
+let
+  fs = lib.fileset;
+  sourceFiles = ./hello.txt;
+in
+
+fs.trace sourceFiles
+
+stdenv.mkDerivation {
+  name = "fileset";
+  src = fs.toSource {
+    root = ./.;
+    fileset = sourceFiles;
+  };
+  postInstall = ''
+    mkdir $out
+    cp -v hello.txt $out
+  '';
+}
+```
+
+The call to `fs.trace` prints the file set that will be used as a derivation input.
+
+The build will succeed (note that it will take some time to fetch the first time around):
```

Comment:
This should take some time already when doing the first `trace`, so either note it there or skip.

```suggestion
Try building it:
```

Reply:
Happens here due to niv not being able to re-use the download.

Changed it to an admonition though (see commit)